Automatic buildcache push after installing from source

rjanalik opened the pull request:
I propose adding a feature that allows users to automatically push to the buildcache after installation of a package, but only if the package is compiled from source.
The key functionalities of this feature include:
- Providing an option to flag a mirror for automatic push when adding a mirror.
- After a package is installed from source, identifying all flagged mirrors.
- Assembling a buildcache push command and pushing the package to the flagged mirrors.

rjanalik commented:
@spackbot fix style

wdconinc commented:
Looks great! We have found that there are benefits to pushing some packages as soon as they are built (robustness in case the build fails later) but for some buildcaches (OCI) it is better to push entire environments only at the end of the build. Do you see a way of achieving that with this autopush strategy?

rjanalik commented:
I was thinking about something like that and I don't know what is a reasonable solution for all possible use cases. For some flags (e.g., --fail-fast) it makes sense to use the same as are used in the install command. I was thinking about using --only for the use case you describe, but I think that would cause only problems.

However, I could add new flag to the install command. For example --push-only={package,dependencies} with package being the default.

rjanalik commented:
But I think for the use case you describe the autopush does not make sense. If you want to install everything and then create an image with the complete environment, I don't see any benefit from this feature.

kwryankrattiger commented on lib/spack/spack/installer.py:
```diff
@@ -1702,6 +1704,29 @@ def _install_task(self, task: BuildTask, install_status: InstallStatus) -> None:
+                # No need to add them again (especially with --force)
+                bc_flags.append("--only=package")
+                # We want the pushed packages to be available right away
+                bc_flags.append("--update-index")
```
`--update-index` and `--force` probably should no be in there. 

`--update-index` is expensive, not necessarily required, and this also should probably consider updating the signing key index. I think two things for this could be to
  a) move the update index part of this to the post install step that loops the `autopush` mirrors after installing all of the packages
  b) make updating the index an option that is passed.

`--force` just seem strange here. The reason in the comments is not possible, if there is a corrupted item in the build cache Spack does not fall back to installing from source, it simply fails the install.

haampie commented on lib/spack/spack/installer.py:
```diff
@@ -1702,6 +1704,29 @@ def _install_task(self, task: BuildTask, install_status: InstallStatus) -> None:
             if task.compiler:
                 self._add_compiler_package_to_config(pkg)
+
+            # Iterate over all mirrors with autopush==True
```
I guess you're putting this here instead of in a post-install hook because it relies on the package being installed in the database, correct?

I'd rather see this in a new post-install hook, cause the installer is already doing (too) many things.

(I've created an issue for `pkg.windows_establish_runtime_linkage()` which certainly does not belong there: https://github.com/spack/spack/issues/42445, similarly the `if task.compiler` bits do not belong there. I guess they've set precedent for this ad-hoc logic, but better not to continue with that...)

Sounds like we need yet another hook that runs after a package is added to the database.

I think we can actually repurpose

```python
spack.hooks.on_install_success(self.pkg.spec)
```

for this, cause as far as I can see, nobody implements that hook anyways

rjanalik replied:
I am putting it in the installer because of my limited knowledge of Spack (I did not know about the hooks). And yes, as it relies on the package being installed (compiled) it seemed logical to put it there.

Thanks for your suggestions, I am going to look at the hooks.

haampie replied:
Perfect! You can also reach out on Slack if you have questions like why or where things are done.

haampie reviewed (changes requested):
I'm OK with the feature, but better to implement it as a hook so that `installer.py` doesn't get more complicated.

Also when it is a hook, it can be unit tested: check if it triggers a push command with the expected flags.

I kinda wish though there was a layer between ArgParse and the actual command, e.g. if buildcache push would do flag validation, and transform `args` as `BuildCacheArgs(mirror=x, force=..., ...)` or so, you could use the latter object here instead.

I'm also reasonably sure people want control over _how_ packages are pushed (see @kwryankrattiger and @wdconinc's comments) , but in principle the `autopush` prop can be extended later as `union[bool, dict]`, so that is fine.

haampie commented:
And regarding (runnable) OCI images / --only=package:

I guess the primary point of autopush is to immediately backup binaries in ephemeral CI environments, so you don't loose them if a later build fails.

Sounds like `buildcache push --only=package` is the thing you want then.

Maybe @wdconinc's issue is mostly (?) addressed when OCI images support `--only=package`, cause later you can create a runnable container image from an environment as a whole with `spack -e my_env buildcache push --tag my_tag`, which contains all roots and runtime dependencies, and should not trigger re-upload of dependencies.

rjanalik commented:
@spackbot fix style

rjanalik commented:
@spackbot fix style

haampie commented on lib/spack/spack/hooks/autopush.py:
```diff
@@ -0,0 +1,39 @@
+            # No need to add them again (especially with --force)
+            bc_flags.append("--only=package")
+            # We want the pushed packages to be available right away
+            bc_flags.append("--update-index")
```
I don't think `--update-index` should be a hard-coded default. `--update-index` is a linear time operation so if you do it for every push it becomes quadratic.

rjanalik replied:
I moved the update of the index until the end of installation, so it is done only once. I created a new hook for that.
What do you think? Do you have an idea how to name that hook?

haampie commented on lib/spack/spack/hooks/autopush.py:
```diff
@@ -0,0 +1,39 @@
+        autopush_mirrors = [mirror for mirror in all_mirrors.values() if mirror.get_autopush()]
+        for mirror in autopush_mirrors:
+            # Create and execute buildcache push command
+            bc_parser = argparse.ArgumentParser()
```
Can you use

```
spack.binary_distribution.push_or_raise
```

instead of `spack.cmd.buildcache.push_fn`?

It doesn't support OCI stuff, but that's fine, cause that doesn't work with `--only=package` anyways.

rjanalik replied:
Done. Now the code look much better.

haampie commented on lib/spack/spack/hooks/autopush.py:
```diff
@@ -0,0 +1,39 @@
+def post_install_in_db(spec):
+    pkg = spec.package
+    # Package was installed from source
+    if not pkg.installed_from_binary_cache:
```
nitpick about style:
```suggestion
    if pkg.installed_from_binary_cache:
        return
    # push
```

haampie commented on lib/spack/spack/hooks/autopush.py:
```diff
@@ -0,0 +1,60 @@
+def post_install(spec, explicit):
+    # Push package to all buildcaches with autopush==True
+
+    # Do nothing if package was not installed from source
+    pkg = spec.package
+
```
```suggestion
```

haampie commented:
Hi @rjanalik, I would leave the "update index" bits out for this PR, could do that as a follow up. Then the scope of this PR is just getting autopush in.

rjanalik commented:
Yes, I agree. I will remove it and let you know. Once the "update index" is removed, the only hook needed is `post_install()`.

rjanalik committed:
Add new feature: buildcache push after installing from source

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

rjanalik commented:
@haampie Done, I removed the update index stuff and squashed the commits.
Should I also remove the first commit dealing with circular imports? It was needed before I moved the implementation to the hook. Now it is not needed, so I suggest moving this refactoring to another PR or completely removing it.

haampie commented:
Looks pretty good, I just quickly glanced over it.

Can you also add documentation for this feature? Maybe after `Relocation` here: https://spack.readthedocs.io/en/latest/binary_caches.html

rjanalik committed:
Add documentation

rjanalik commented:
Documentation done. What do you think?

rjanalik committed:
Add autopush to miror set subcommand

rjanalik committed:
Add test: mirror add set autopush

rjanalik committed:
Add test: install autopush

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -148,7 +148,7 @@ def push_url(self):
 
     def _update_connection_dict(self, current_data: dict, new_data: dict, top_level: bool):
-        keys = ["url", "access_pair", "access_token", "profile", "endpoint_url"]
+        keys = ["url", "access_pair", "access_token", "profile", "endpoint_url", "autopush"]
```
```suggestion
        keys = ["url", "access_pair", "access_token", "profile", "endpoint_url"]
```

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -148,7 +148,7 @@ def push_url(self):
-        keys = ["url", "access_pair", "access_token", "profile", "endpoint_url"]
+        keys = ["url", "access_pair", "access_token", "profile", "endpoint_url", "autopush"]
         if top_level:
             keys += ["binary", "source", "signed"]
```
```suggestion
            keys += ["binary", "source", "signed", "autopush"]
```

haampie committed:
Simplify test; make autopush a prop & only top-level

haampie reviewed (approved):
Thanks!
